### PR TITLE
Add code to scrape information from YouTube

### PR DESCRIPTION
**Pull request by witchdocxxx**
1. Added YouTube scraper that scrapes the whole "Music in this video" section (if available) or gets hold of the Artist and Song name needed to input to Spotify's API.

2. Added helpers.py that helps in getting "Artist" name if the title of the YT video is only "Song". it also changes shortened link to long link for convenience.

3. Fixed regex for shortened links in validator.py and also took care that songs embedded in playlist are not validated.

### CODE REVIEW

**Comment by witchdocxxx** [comment]
@ghostwriternr @mukul-mehta @thescriptninja please review.

File: extractor/youtube/core.py [hunk 1]
```diff
@@ -1,10 +1,48 @@
...
+    else :
+        title = get_artist_title(raw_title)
+        artist = soup.find('a',class_= "yt-uix-sessionlink spf-link").text
+        artist = clean_channel(artist)
+
+    info = {'Category': None , 'Song' :title, 'Artist':artist, 'Album':None, 'Licensed to YouTube by' : None}
+    return info
+
+# Scrapper function for videos with "Music in this video" section
+def scrape_yt_new(soup):
```
Comment (ghostwriternr) [inline extractor/youtube/core.py:26]: Can we explicitly name the function to say what it is doing? Something like `scrape_embedded_yt_metadata`.

File: extractor/youtube/core.py [hunk 2]
```diff
@@ -1,10 +1,48 @@
...
+        raw_title =  get_artist_title(raw_title)
+        artist, title = raw_title.split(' - ')
+
+    # In case the  YouTube Title only contains song name    
+    else :
+        title = get_artist_title(raw_title)
+        artist = soup.find('a',class_= "yt-uix-sessionlink spf-link").text
+        artist = clean_channel(artist)
+
+    info = {'Category': None , 'Song' :title, 'Artist':artist, 'Album':None, 'Licensed to YouTube by' : None}
```
Comment (ghostwriternr) [inline extractor/youtube/core.py:22]: Why an `info` dictionary? Can we instead create a class and return an instance of that?

File: extractor/youtube/core.py [hunk 3]
```diff
@@ -1,10 +1,48 @@
...
+
+    #In case the YouTube Title is in good format,i.e, Artist-Song("foo")
+    if '-' in raw_title:
+        raw_title =  get_artist_title(raw_title)
+        artist, title = raw_title.split(' - ')
+
+    # In case the  YouTube Title only contains song name    
+    else :
+        title = get_artist_title(raw_title)
+        artist = soup.find('a',class_= "yt-uix-sessionlink spf-link").text
```
Comment (ghostwriternr) [inline extractor/youtube/core.py:19]: Can you explain (as a reply to this) what this line is doing? I'm not sure I understand why we're treating titles with `-` separately.

File: extractor/youtube/validator.py [hunk 4]
```diff
@@ -1,7 +1,11 @@
...
-    return False
+def yt_validate(yt_link) :
+        if(bool(re.match("^https://\w{0,3}.?youtube+\.\w{2,3}/watch\?v=[\w-]{11}$", yt_link)) or bool(re.match("^https://youtu.be/", yt_link))):
+                return True
+        return False
 
+def is_short_link(yt_link):
+        if(bool(re.match("^https://youtu.be/", yt_link))):
+                return True
+        return False
```
Comment (ghostwriternr) [inline extractor/youtube/validator.py:11]: I'm not sure if this covers all variations. Please validate against the expressions in this link: https://stackoverflow.com/a/3726073/4396392.

File: extractor/youtube/core.py (hunk 3 above)
Comment (witchdocxxx) [reply extractor/youtube/core.py:19]: Most of the youtube music video titles have format as "Artist - Song". So I used it.

File: extractor/youtube/validator.py (hunk 4 above)
Comment (witchdocxxx) [reply extractor/youtube/validator.py:11]: Sure I'll look into it.

**Comment by thescriptninja** [comment]
@ghostwriternr please review. I added a single entrypoint like you said.
```
from extractor.core import SongData
data = SongData("<YouTube-URL>")
```

File: extractor/youtube/core.py (hunk 3 above)
Comment (witchdocxxx) [reply extractor/youtube/core.py]: Corrected it.

File: extractor/youtube/validator.py [hunk 5]
```diff
@@ -1,7 +1,18 @@
...
+        if(bool(re.match("http(?:s?):\/\/(?:www\.)?youtu(?:be\.com\/watch\?v=|\.be\/)([\w\-\_]*)(&(amp;)?‌​[\w\?‌​=]*)?", yt_link))):
+                return True
+        return False
+
+def available_video(soup):
+        if(soup.find('meta', {'property': 'og:title'})):
+                return True
+        return False
+
+def is_short_link(yt_link):
```
Comment (ghostwriternr) [inline extractor/youtube/validator.py:13]: This method is probably not needed anymore, no? If you look at the regex in yt_validate, it already handles short URLs gracefully.

File: extractor/youtube/helpers.py [hunk 6]
```diff
@@ -0,0 +1,23 @@
+from extractor.youtube.validator import is_short_link
+
+# Get long youtube link from shortened link
+def get_yt_link(link):
+    if is_short_link(link) :
```
Comment (ghostwriternr) [inline extractor/youtube/helpers.py:5]: If `is_short_link` isn't used, we can remove the usage here.

**Comment by thescriptninja** [comment]
Cool! This PR is done. :partying_face:

**Pull request merged**